### PR TITLE
Exceptions instead of yield from / returns

### PR DESCRIPTION
**Pull request by kozzztik**
I`m dont sure, was it decided specially to use yield from / return construction for abnormal method finish or it is migrated from original smtpd. 
May be its better to use exceptions for that, as it is a case they was designed for. It is a part of my #41 vision, where same exception is used is user handlers to indicate errors.
That is only idea, may be there was a reasons for yield from / returns. Anyway, that is more about code style, then functionality.

### CODE REVIEW

**Comment by kozzztik** [comment]
May be better to use name SMTPError instead of SMTPExceptions.

File: aiosmtpd/smtp.py [hunk 1]
```diff
@@ -570,10 +550,7 @@ def smtp_DATA(self, arg):
             else:
                 status = self.event_handler.process_message(*args)
```
Comment (Mortal) [inline aiosmtpd/smtp.py]: The assignment to local variable "status" is not needed here and should be removed.

**Comment by Mortal** [comment]
Initially I thought that using exceptions for SMTP error responses is surprising/unexpected (since I don't think that error responses (4xx, 5xx) in the SMTP session are exceptional circumstances). However, it occurred to me that I am in a way guilty of this myself, using an [InvalidRecipient exception to signal that a 550 response is needed](https://github.com/TK-IT/mail/blob/ac99551dd4adf4c08d6ab0a186a1024328f824d7/emailtunnel/__init__.py#L592-L599).

Allowing user code in RCPT to either return a recipient object or raise an error response is appealing somehow. From my own experience with mailing list servers I can't think of other error responses than 550 Mailbox unavailable -- is that error response common enough to warrant its own subclass of SMTPError in aiosmtpd? What other SMTP error responses could a Handler object be expected to raise?

**Comment by warsaw** [comment]
As I mentioned in #41 my own sensibilities lean away from using exceptions for error code responses, exactly as you state: to me they aren't exceptional cases.  It's entirely reasonable and expected for example, for a non 250 return code to come back from a RCPT TO with an invalid recipient.

I also don't think it's necessary, since everything that `handle_RCPT()` or `handle_MAIL()` wants to do can be done directly on the `envelope.mail_from` or `envelope.rcpt_tos` attributes.

I'm not *as* concerned about performance, but do be aware that raising and catching exceptions are generally more expensive than just yielding a status code.

**Comment by kozzztik** [comment]
Ok, performance matters. That was bad idea.

**Pull request closed without merging**